text/0046-metadata-for-offline-signers.md

```diff
@@ -0,0 +1,295 @@
+
+### Solution requirements
+
+#### Include metadata digest into signature
+
+Some cryptographically strong digest of metadata MAY be included into signable blob. There SHALL NOT be storage overhead for this blob, nor computational overhead; thus MUST be a constant within a given runtime, deterministically defined by metadata.
+
+ - Metadata information that could be used in signable extrinsic decoding MAY be included in digest, its inclusion MUST be indicated in signed extensions;
+ - Digest MUST be deterministic with respect to metadata;
+ - Digest MUST be cryptographically strong against pre-image, both first (finding an input that results in given digest) and second (finding an input that results in same digest as some other input given);
+ - Extra-metadata information necessary for extrinsic decoding and constant within runtime version MUST be included in digest;
+ - It SHOULD be possible to quickly withdraw offline signing mechanism without access to cold signing devices;
+ - Digest format SHOULD be versioned.
+ - Work necessary for proving metadata authenticity MAY be omitted at discretion of signer device design (to support automation tools).
+
+#### Reduce metadata size
+
+Metadata should be stripped from parts that are not necessary to parse a signable extrinsic, then it should be separated into a finite set of self-descriptive chunks. Thus, a subset of chunks necessary for signable extrinsic decoding and rendering could be sent, possibly in small portions (ultimately - one at a time), to cold device together with proof.
+
+ - Single chunk with proof payload size SHOULD fit within few kB;
+ - Chunks handling mechanism SHOULD support chunks being sent in any order without memory utilization overhead;
+ - Unused enum variants MUST be stripped (this has great impact on transmitted metadata size; examples: era enum, enum with all calls for call batching).
+
+## Stakeholders
+
+All chain teams are stakeholders, as implementing this feature would require timely effort on their side and would impact compatibility with older tools.
+
+This feature is essential for **all** offline signer tools; many regular signing tools might make use of it. In general, this RFC greatly improves security of any network implementing it, as many governing keys are used with offline signers.
+
+Implementing this RFC would remove requirement to maintain metadata portals manually, as task of metadata verification would be effectively moved to consensus mechanism of the chain.
+
+## Explanation
+
+Detailed description of metadata shortening and digest process is provided in [metadata-shortener](https://github.com/Alzymologist/metadata-shortener) crate (see `cargo doc --open` and examples). Below are presented algorithms of the process.
+
+### Definitions
+
+
+#### Metadata descriptor
+
+Values for:
+
+1. `u8` metadata shortening protocol version, 
+2. SCALE-encoded `ExtrinsicMetadata`,
+3. SCALE-encoded `spec_version` `String`,
```

Comment:
why this is string, not u32?

Reply:
Because we've seen parachains and standalone chains where this was not `u32`. Those went extinct some time ago though. This is not forbidden anywhere and people are experimenting sometimes, there is no reason to punish them. Or is there? We certainly should change it into `u32` (it would be easier and cleaner indeed) if this is planned to be forced, but that has to be documented well in chain developer docs first, which is out of my reach.